backend/packages/Upgrade/src/api/services/MoocletRewardsService.ts

```diff
@@ -0,0 +1,207 @@
+import { UpgradeLogger } from '../../../src/lib/logger/UpgradeLogger';
+import { REPEATED_MEASURE, IMetricMetaData, ILogInput } from 'upgrade_types';
+import { RequestedExperimentUser } from '../controllers/validators/ExperimentUserValidator';
+import { Metric } from '../models/Metric';
+import { MoocletExperimentRef } from '../models/MoocletExperimentRef';
+import { v4 as uuid } from 'uuid';
+import { MetricService } from './MetricService';
+import { MoocletDataService } from './MoocletDataService';
+import { MoocletExperimentRefRepository } from '../repositories/MoocletExperimentRefRepository';
+import { IndividualEnrollment } from '../models/IndividualEnrollment';
+import { IndividualEnrollmentRepository } from '../repositories/IndividualEnrollmentRepository';
+import { MoocletValueRequestBody } from '../../../src/types/Mooclet';
+import { Service } from 'typedi';
+import { InjectRepository } from '../../typeorm-typedi-extensions';
+import { BinaryRewardMetricAllowedValue, BinaryRewardMetricValueMap } from '../../../../../../types/src/Mooclet';
+
+interface FindRewardParams {
+  user: RequestedExperimentUser;
+  moocletExperimentRefs: MoocletExperimentRef[];
+  logger: UpgradeLogger;
+}
+
+interface SyncRewardParams {
+  moocletExperimentRef: MoocletExperimentRef;
+  metricValue: string | number;
+  logger: UpgradeLogger;
+}
+
+@Service()
+export class MoocletRewardsService {
+  constructor(
+    @InjectRepository()
+    private moocletExperimentRefRepository: MoocletExperimentRefRepository,
+    @InjectRepository()
+    private individualEnrollmentRepository: IndividualEnrollmentRepository,
+    private metricService: MetricService,
+    private moocletDataService: MoocletDataService
+  ) {}
+
+  public attachRewardMetricQuery(rewardMetricKey: string, queries: any[]) {
+    queries.push({
+      id: uuid(),
+      name: 'Success Rate',
+      query: {
+        operationType: 'percentage',
+        compareFn: '=',
+        compareValue: BinaryRewardMetricAllowedValue.SUCCESS,
+      },
+      metric: {
+        key: rewardMetricKey,
+      },
+      repeatedMeasure: REPEATED_MEASURE.mostRecent,
+    });
+  }
+
+  public async createAndSaveRewardMetric(
+    rewardMetricKey: string,
+    context: string,
+    logger: UpgradeLogger
+  ): Promise<Metric[]> {
+    const metric = {
+      id: uuid(),
+      metric: rewardMetricKey,
+      datatype: IMetricMetaData.CATEGORICAL,
+      allowedValues: [BinaryRewardMetricAllowedValue.SUCCESS, BinaryRewardMetricAllowedValue.FAILURE],
+    };
+
+    return this.metricService.saveAllMetrics([metric], [context], logger);
+  }
+
+  public async parseLogsForActiveRewardMetricKey(
```

Comment:
The code is jumping from one function to another instead, which is making it difficult to follow can we have all the control in a single function, and all the modular functions should return the result after doing specific action to the control function?

```
// Get all active mooclet experiment

// Get log attributes

// Find experiment with reward metrics keys which exists in the log attributes

// Determine versions for rewards

// Send rewards
```

Reply:
yes i can do that